Replace assertions with verifies in ocean.io.*

nemanja-boric-sociomantic opened the pull request.

nemanja-boric-sociomantic committed:
Remove debug (test) programs

These were never compiled nor maintained.

nemanja-boric-sociomantic committed:
Change asserts to verifies

nemanja-boric-sociomantic committed:
Use test instead of assert in unittest

nemanja-boric-sociomantic commented on src/ocean/io/select/client/FiberSelectEvent.d:
```diff
@@ -48,8 +48,12 @@ module ocean.io.select.client.FiberSelectEvent;
+
 import ocean.core.MessageFiber;
 
+import ocean.core.Verify;
```
double import

nemanja-boric-sociomantic commented on src/ocean/io/select/client/FileSystemEvent.d:
```diff
@@ -383,10 +384,10 @@ class FileSystemEvent : ISelectClient
         foreach ( ev; this.fd.readEvents() )
         {
-            assert(ev.mask);
+            verify(ev.mask != 0);
```
Probably would be better off with `ev.mask != ev.mask.init`

nemanja-boric-sociomantic commented on src/ocean/io/select/client/SelectEvent.d:
```diff
@@ -251,6 +248,7 @@ public class SelectEvent : ISelectEvent
 
     protected override bool handle_ ( ulong n )
     {
+        verify(this.handler !is null);
```
this one is added here, because it was in line 147, but never compiled (in contracts of the abstract method's implementations are not compiled).

nemanja-boric-sociomantic commented on src/ocean/io/stream/Data.d:
```diff
@@ -450,7 +452,7 @@ class DataOutput : OutputFilter
         {
                 auto count = output.write (src[0..bytes]);
-                assert (count is bytes);
+                verify(count == bytes);
```
Is there any difference in using `is` vs `==` for value types?

mihails-strasuns-sociomantic replied:
For structs only (as they may provide own `opEqual`)

nemanja-boric-sociomantic commented on src/ocean/io/stream/Iterator.d:
```diff
@@ -58,6 +60,7 @@ package import ocean.io.device.Conduit : InputFilter, InputBuffer, InputStream;
 
 class Iterator(T) : InputFilter
 {
+        import ocean.core.Verify;
```
double import

gavin-norman-sociomantic reviewed (commented):
All looks ok to me. I agree with your comments.

david-eckardt-sociomantic commented on src/ocean/io/select/protocol/task/TaskSelectClient.d:
```diff
@@ -161,9 +160,9 @@ class TaskSelectClient: ISelectClient
         {
             this.task = task.getThis();
-            assert(this.task);
+            verify(this.task !is null);
```
In cases like this we have to be careful: `verify(object !is null)` does not call the object invariant while `assert(object)` does. `Task` doesn't have an invariant so it's fine here, but in general this change can silently drop a debugging aid.

gavin-norman-sociomantic replied:
So the better thing to do would be:
```D
// old
assert(object);

// replace with
verify(object !is null && object); // call invariant
```

david-eckardt-sociomantic replied:
No, only literally `assert(object)` works. It is a magic expression to call the invariant.

mihails-strasuns-sociomantic replied:
And `assert(task)` is perfectly fine here:
1) it will just crash on the next line in release build
2) it protects against being called from main thread context - not something that may not be noticed before production deloyment

david-eckardt-sociomantic commented on src/ocean/io/serialize/StructSerializer.d:
```diff
@@ -206,7 +200,7 @@ struct StructSerializer ( bool AllowUnions = false )
     in
     {
-        assert (s, typeof (*this).stringof ~ ".transmit (receive = " ~
+        verify(s, typeof (*this).stringof ~ ".transmit (receive = " ~
```
Should this be `s !is null`?

nemanja-boric-sociomantic replied:
There's an `in` contract here anyway :-)

nemanja-boric-sociomantic commented on src/ocean/io/Path.d:
```diff
@@ -1460,11 +1462,11 @@ in
             switch (pattern[i])
                    {
                    case '[':
-                        assert(!inbracket);
```
I've grepped for `in$` and this showed up as a remaining `in` contract. This feels very expensive to be left in `release` build. Should I just leave it here and use `assert`?

gavin-norman-sociomantic replied:
Is `patternMatch` a function that you could imagine being called in time critical code?

nemanja-boric-sociomantic replied:
Puh, I guess you're right. It's still `O(n)`, but let's KISS.

mihails-strasuns-sociomantic replied:
I would keep it nested inside `{}` block to make it easier to eventually reconsider and remove it if needed

nemanja-boric-sociomantic committed:
Move debug block from in contract to body

nemanja-boric-sociomantic commented:
Updated to address @david-eckardt-sociomantic 's comment and to move one debug block from `in` contract (it couldn't fail, it was just printout) to the `body`.

nemanja-boric-sociomantic commented:
Updated.

mihails-strasuns-sociomantic commented on src/ocean/io/FilePath.d:
```diff
@@ -986,12 +987,9 @@ class FilePath : PathView
-    }
-    body
     {
+        verify(dst !is null);
```
`verify(dst.length > 0)` makes more sense here

mihails-strasuns-sociomantic commented on src/ocean/io/model/SuspendableThrottlerCount.d:
```diff
@@ -33,6 +33,7 @@ module ocean.io.model.SuspendableThrottlerCount;
 *******************************************************************************/
 
 import ocean.transition;
+import ocean.core.Verify;
```
I think all checks in this module actually should remain `assert` - they don't protect from any kind of memory corruption and relate to hot path in apps.

gavin-norman-sociomantic replied:
>they don't protect from any kind of memory corruption

But they are valid sanity checks, which is the purpose of `verify`. I don't see any harm leaving them as `verify`.

nemanja-boric-sociomantic replied:
This is what I think as well. Perhaps the trick is "hot path"?

mihails-strasuns-sociomantic replied:
No, purpose of `verify` is specifically sanity checking you _can not abandon_. So that when you build `-release` you get maximum performance possible while still not compromising the app. No more, no less.

mihails-strasuns-sociomantic replied:
In other words, "sanity check that is nice to have to catch weird app state but which won't help to keep the app working and won't prevent anything bad" is pretty much definition of `assert` use case.

gavin-norman-sociomantic replied:
Huh... this makes it a very tricky thing to decide, in many cases. A bug in this class, for example, could lead to seriously wrong app behaviour. Can those checks be abandoned?

Also: I was viewing this conversion process as allowing us to run release builds of D2 apps without losing any of the sanity checks that are in place right now but also without the risk of `Error`s being thrown. This is a somewhat different (and I would argue far more pressing) goal.

mihails-strasuns-sociomantic commented on src/ocean/io/select/client/EpollProcess.d:
```diff
@@ -300,8 +302,8 @@ public abstract class EpollProcess
                 {
-                    assert( errno() == ECHILD );
-                    assert( this.processes.length == 0 );
+                    verify(errno() == ECHILD);
```
either `assert` or `enforce` - either this is protecting from OS being broken or checking external state that is beyond programmer control

mihails-strasuns-sociomantic commented on src/ocean/io/select/client/FiberSelectEvent.d:
```diff
@@ -187,12 +189,9 @@ public class FiberSelectEvent : IFiberSelectClient
-    }
-    body
     {
+        verify(super.fiber.waiting);
```
`assert` is fine, fiber will have own `verify` on resume

mihails-strasuns-sociomantic commented on src/ocean/io/select/client/FiberTimerEvent.d:
```diff
@@ -139,12 +141,8 @@ public class FiberTimerEvent : IFiberSelectClient
-    }
-    body
     {
+        verify(s >= 0); // tolerate +∞
```
Does that check even work as commented with floats?

nemanja-boric-sociomantic replied:
Yeah, that works.

mihails-strasuns-sociomantic commented on src/ocean/io/select/client/FiberTimerEvent.d:
```diff
@@ -188,12 +186,8 @@ public class FiberTimerEvent : IFiberSelectClient
-    }
-    body
     {
+        verify(this.fiber.waiting);
```
Same comment, redundant

mihails-strasuns-sociomantic commented on src/ocean/io/select/protocol/fiber/model/IFiberSelectProtocol.d:
```diff
@@ -236,12 +235,9 @@ abstract class IFiberSelectProtocol : IFiberSelectClient
-    }
-    body
     {
+        verify(this.fiber.waiting);
```
Same comment, redundant

mihails-strasuns-sociomantic commented on src/ocean/io/select/protocol/fiber/model/IFiberSelectProtocol.d:
```diff
@@ -268,12 +264,9 @@ abstract class IFiberSelectProtocol : IFiberSelectClient
-    }
-    body
     {
+        verify(this.fiber.running);
```
Same

mihails-strasuns-sociomantic commented on src/ocean/io/select/protocol/task/TaskSelectClient.d:
```diff
@@ -127,17 +128,15 @@ class TaskSelectClient: ISelectClient
     body
     {
+        verify(events_expected != Event.init);
+        verify(this.task is null);
```
At least latter was fine as `assert`

nemanja-boric-sociomantic replied:
Ah you're right.

mihails-strasuns-sociomantic commented on src/ocean/io/select/protocol/task/TaskSelectClient.d:
```diff
@@ -161,9 +160,9 @@ class TaskSelectClient: ISelectClient
+            verify(this.task !is null);
             this.task.suspend();
-            assert(!this.task);
+            verify(this.task is null);
```
Also can be `assert`

mihails-strasuns-sociomantic commented on src/ocean/io/serialize/StringStructSerializer.d:
```diff
@@ -219,7 +220,7 @@ public class StringStructSerializer ( Char )
     public void open ( ref Char[] output, cstring name )
     {
-        assert(this.indent.length == 0, "Non-zero indentation in open");
+        verify(this.indent.length == 0, "Non-zero indentation in open");
```
1) Error message is reverse wrong btw.
2) I think all those check must be `assert` (probably with reusable exception even) as serialization often happens with I/O data

nemanja-boric-sociomantic replied:
It doesn't look that the error message is wrong :)

mihails-strasuns-sociomantic replied:
So many negatives :grumpy:

mihails-strasuns-sociomantic commented on src/ocean/io/serialize/StructSerializer.d:
```diff
@@ -125,12 +127,8 @@ struct StructSerializer ( bool AllowUnions = false )
-    }
-    body
     {
+        verifyStructPtr!("dump")(s);
```
Same for this module - it all should `enforce` with actual exception, those are not sanity violations

mihails-strasuns-sociomantic reviewed (changes requested):
Some arguable comments from my side :)

mihails-strasuns-sociomantic commented:
With last discussion in mind only remaining thing I'd like to happen is changing `verify` to `enforce` in serializer

nemanja-boric-sociomantic commented:
I'll change them :+1:

nemanja-boric-sociomantic committed:
Use enforce for verifications in ocean.io.serialize.*StructSerializer

Most serialization is being done with the external data, so in this case
it's more appropriate to do enforce in this cases, with the reusable
exception.

nemanja-boric-sociomantic commented:
Updated with the enforcements for the `StringStructSerializer` and `StructSerializer`.

mihails-strasuns-sociomantic commented:
Red

```
./src/ocean/io/serialize/StructSerializer.d(64): Error: constructor ocean.io.serialize.StructSerializer.SerializerException.ReusableExceptionImplementation!().this (ulong size = cast(ulong)128) does not match parameter types (char[],char[],int)
./src/ocean/io/serialize/StructSerializer.d(64): Error: expected 1 arguments, not 3 for non-variadic function type SerializerException(ulong size = cast(ulong)128)
./src/ocean/io/serialize/StructSerializer.d(64): Error: cannot implicitly convert expression (msg) of type char[] to ulong
```

nemanja-boric-sociomantic commented:
Updated.
